be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -43,65 +44,60 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
   }
 
   // This function will try to digest the set of messages for an entire channel
-  // It will go channel by channel until it finished processing messages or MAX_RETRY is reached
+  // It will go channel by channel until it finished processing messages or MAX_RETRY_PER_MESSAGE * total_number_of_message is reached
   @tailrec
   private def passThroughPipeline(
-      receivedResponse: Map[Channel, Set[GraphMessage]],
-      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed],
-      remainingAttempts: Int
+      receivedResponse: List[(GraphMessage, Int)],
+      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed]
   )(implicit system: ActorSystem): Boolean = {
-    if (receivedResponse.isEmpty || remainingAttempts <= 0) {
+    if (receivedResponse.isEmpty || receivedResponse.forall(elem => elem._2 <= 0)) {
       return receivedResponse.isEmpty
     }
 
-    var failedMessages: Map[Channel, Set[GraphMessage]] = Map.empty
+    var failedMessages: List[(GraphMessage, Int)] = Nil
     receivedResponse.foreach {
-      case (channel, messagesSet) =>
-        val failedMsgErrorMap = messagesThroughPipeline(validatorFlow, messagesSet)
-        if (failedMsgErrorMap.nonEmpty) {
-          failedMessages += channel -> failedMsgErrorMap.keySet
-
-          // only log a during last attempt
-          if (remainingAttempts == 1) {
-            println("Errors: " + failedMsgErrorMap.map {
-              case (msg, error) => error.toString + "\n On:\n" + prettyPrinter(msg)
-            })
+      case (graphMessage, retry) =>
+        if (retry > 0) {
+          messagesThroughPipeline(validatorFlow, graphMessage) match {
+            case Left(err) =>
+              failedMessages = (graphMessage, retry - 1) :: failedMessages
+              // only log a during last attempt
+              if (retry == 1) {
+                println("Errors: " + err.toString + "\n On:\n" + prettyPrinter(graphMessage))
+              }
+            case _ => /* DO NOTHING */
           }
         }
     }
 
-    passThroughPipeline(failedMessages, validatorFlow, remainingAttempts - 1)
+    passThroughPipeline(failedMessages.reverse, validatorFlow)
```

Comment:
Why do you need to reverse your list of messages ?

Reply:
I unfortunately don't have any data to back this up but:
Since we are ordering the channels based on the length, I  wanted to keep their order. (since the we add the `graphMessage` on the top of the failed list, the order is reversed) 

I don't know performance wise if one way is better than the other though...

Reply:
Is there really a need to keep that ordering all the way through the simulation pipeline?

Reply:
I removed since its utility beside being confusing is questionable.